src/bit-systems/waypoint.ts

```diff
@@ -0,0 +1,209 @@
+    takeOwnership(world, spawnPoint);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+function* trySpawnIntoOccupiable(world: HubsWorld, characterController: CharacterControllerSystem) {
+  for (let i = 0; i < 3; i++) {
+    const spawnPoints = occupiableSpawnPoints(world);
+    if (!spawnPoints.length) return false;
+
+    const waypoint = spawnPoints[Math.floor(Math.random() * spawnPoints.length)];
+    if (yield* tryOccupyAndSpawn(world, characterController, waypoint)) return true;
+  }
+
+  return false;
+}
+
+function* moveToSpawnPointJob(world: HubsWorld, characterController: CharacterControllerSystem) {
+  if (yield* trySpawnIntoOccupiable(world, characterController)) return;
+
+  const spawnPoints = nonOccupiableSpawnPoints(world);
+  if (spawnPoints.length) {
+    const waypoint = spawnPoints[Math.floor(Math.random() * spawnPoints.length)];
+    moveToWaypoint(world, waypoint, characterController, true);
+  } else {
+    console.warn("Could not find any available spawn points, spawning at the origin.");
+    characterController.enqueueWaypointTravelTo(new Matrix4().identity(), true, {
+      willDisableMotion: false,
+      willDisableTeleporting: false,
+      snapToNavMesh: true,
+      willMaintainInitialOrientation: false
+    });
+  }
+}
+
+let spawnJob: Coroutine | null = null;
+export function moveToSpawnPoint(world: HubsWorld, characterController: CharacterControllerSystem) {
+  spawnJob = coroutine(moveToSpawnPointJob(world, characterController));
+}
+
+function moveToWaypoint(
+  world: HubsWorld,
+  eid: number,
+  characterController: CharacterControllerSystem,
+  instant: boolean
+) {
+  const obj = world.eid2obj.get(eid)!;
+  obj.updateMatrices();
+
+  characterController.enqueueWaypointTravelTo(
+    obj.matrixWorld,
+    instant || !window.APP.store.state.preferences.animateWaypointTransitions, // TODO: Use store-instance
+    {
+      willDisableMotion: !!(Waypoint.flags[eid] & WaypointFlags.willDisableMotion),
+      willDisableTeleporting: !!(Waypoint.flags[eid] & WaypointFlags.willDisableTeleporting),
+      snapToNavMesh: !!(Waypoint.flags[eid] & WaypointFlags.snapToNavMesh),
+      willMaintainInitialOrientation: !!(Waypoint.flags[eid] & WaypointFlags.willMaintainInitialOrientation)
+    }
+  );
+}
+
+const hoveredLeftWaypointQuery = defineQuery([Waypoint, HoveredRemoteLeft]);
+const hoveredRightWaypointQuery = defineQuery([Waypoint, HoveredRemoteRight]);
+
+const exitedOwnedQuery = exitQuery(defineQuery([Owned]));
+
+let preview: Object3D | null;
+export function waypointSystem(
+  world: HubsWorld,
+  characterController: CharacterControllerSystem,
+  sceneIsFrozen: boolean
+) {
+  if (exitedOwnedQuery(world).includes(myOccupiedWaypoint)) {
+    myOccupiedWaypoint = 0;
+  }
+
+  waypointQuery(world).forEach(eid => {
+    if (hasComponent(world, NetworkedWaypoint, eid) && hasComponent(world, Owned, eid) && eid !== myOccupiedWaypoint) {
```

Comment:
Not sure its avoidable but noting it anyway.. Its still a bit awkward that each system will have to handle this "inherited" case (though maybe most don't need to do anything special?).

Reply:
(From our conversation) One way to avoid this special case here would be to store _who_ occupies the waypoint rather than just _whether_ it's occupied, and then checking if `NetworkedWaypoint.occupied[eid] === Networked.owner[eid]`.